File: backend/app/api/brand_library.py
import uuid

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.auth import get_current_user
from app.core.database import get_db
from app.models.user import User
from app.schemas.brand_library import (
    BrandLibraryListResponse,
    BrandLibraryDetailResponse,
    BrandLibraryItemResponse,
    PublishToLibraryRequest,
    ReviewLibraryItemRequest,
    RemixResponse,
)
from app.services.brand_library_service import (
    list_library_items,
    get_library_item_detail,
    publish_to_library,
    review_library_item,
    remix_library_item,
)

router = APIRouter(prefix="/api/brand-library", tags=["brand-library"])


@router.get("", response_model=BrandLibraryListResponse)
async def get_library_items(
    search: str | None = Query(None),
    product: str | None = Query(None),
    page: int = Query(1, ge=1),
    per_page: int = Query(20, ge=1, le=100),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> BrandLibraryListResponse:
    items, total = await list_library_items(
        db, current_user, search=search, product=product, page=page, per_page=per_page
    )
    return BrandLibraryListResponse(items=items, total=total, page=page, per_page=per_page)


@router.get("/{item_id}", response_model=BrandLibraryDetailResponse)
async def get_library_item(
    item_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> dict:
    return await get_library_item_detail(db, item_id, current_user)


@router.post("", response_model=BrandLibraryItemResponse, status_code=201)
async def publish_artifact(
    data: PublishToLibraryRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> dict:
    item = await publish_to_library(db, current_user, data.artifact_id)
    # Re-fetch with relationships for response
    return await get_library_item_detail(db, item.id, current_user)


@router.patch("/{item_id}", response_model=BrandLibraryDetailResponse)
async def review_item(
    item_id: uuid.UUID,
    data: ReviewLibraryItemRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> dict:
    await review_library_item(db, current_user, item_id, data.action, data.reason)
    return await get_library_item_detail(db, item_id, current_user)


@router.post("/{item_id}/remix", response_model=RemixResponse, status_code=201)
async def remix_item(
    item_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> RemixResponse:
    project_id, artifact_id = await remix_library_item(db, current_user, item_id)
    return RemixResponse(project_id=project_id, artifact_id=artifact_id)

File: frontend/src/lib/api/brand-library.ts
import { apiClient } from "@/lib/api-client";
import type { BrandLibraryListResponse, BrandLibraryItem } from "@/types/brand-library";

export interface BrandLibraryDetailItem extends BrandLibraryItem {
  artifact: BrandLibraryItem["artifact"] & {
    content: Record<string, unknown> | null;
    compliance_score: number | null;
  };
  rejection_reason: string | null;
}

export async function fetchLibraryItems(
  filters?: { search?: string; product?: string },
  page = 1,
  perPage = 20
): Promise<BrandLibraryListResponse> {
  const params = new URLSearchParams();
  if (filters?.search) params.set("search", filters.search);
  if (filters?.product) params.set("product", filters.product);
  params.set("page", String(page));
  params.set("per_page", String(perPage));
  return apiClient.get<BrandLibraryListResponse>(`/api/brand-library?${params}`);
}

export async function fetchLibraryItemDetail(
  id: string
): Promise<BrandLibraryDetailItem> {
  return apiClient.get<BrandLibraryDetailItem>(`/api/brand-library/${id}`);
}

export async function publishToLibrary(
  artifactId: string
): Promise<BrandLibraryDetailItem> {
  return apiClient.post<BrandLibraryDetailItem>("/api/brand-library", {
    artifact_id: artifactId,
  });
}

export async function reviewLibraryItem(
  id: string,
  action: "approve" | "reject" | "unpublish",
  reason?: string
): Promise<BrandLibraryDetailItem> {
  return apiClient.patch<BrandLibraryDetailItem>(`/api/brand-library/${id}`, {
    action,
    reason,
  });
}

export async function remixLibraryItem(
  id: string
): Promise<{ project_id: string; artifact_id: string }> {
  return apiClient.post<{ project_id: string; artifact_id: string }>(
    `/api/brand-library/${id}/remix`
  );
}
